scipy/special/cephes/stirling.c

```diff
@@ -0,0 +1,54 @@
+/* c implementation of Stirling numbers of the second kind */
+
+/*
+ *
+ *     Stirling numbers of the second kind
+ *
+ *
+ *
+ * SYNOPSIS: Stirling numbers of the second kind count the
+ *  number of ways to make a partition of n distinct elements
+ *  into k non-empty subsets.
+ *
+ * DESCRIPTION: n is the number of distinct elements of the set
+ *  to be partitioned and k is the number of non-empty subsets.
+ *  The values for n < 0 or k < 0 are interpreted as 0. If you
+ *
+ * ACCURACY: The method returns an unsigned integer type and
+ *  the size of the type 32 bits.
+ *
+ *
+ * NOTE: this file is *NOT* part of the ceph distribution and was
+ *  added by Lucas Roberts in 2023 after opening this issue
+ *  https://github.com/scipy/scipy/issues/17890
+ */
+#include "mconf.h"
+
+int stirling2(int n, int k){
+    if (k < 0 || k > n || n < 0){
+        return 0;
+    }
+    int arraySize = n + 1;
+    int prev[arraySize];
+    int curr[arraySize];
```

Comment:
> Also, by declaring them on the stack, we can blow up the stack and get a segfault even with sane input. On my 64bit Linux machine I found that
> 
> ```python
> import scipy.special as sc
> 
> stirling2(2000000, 1999999)
> # Segmentation fault (core dumped)
> ```

Ok. FWIW the calculation will overflow long before this point so mayvbe it makes sense to raise an error if `n > X` where `X` is the value of overflow (17 on my machine w/`int` being 32 bit IIRc)? 

>We should calloc these instead of declaring them on the stack (making sure to call free before returning). VLAs are part of the C99 standard, but became optional as of C11. I checked the [toolchain roadmap](https://docs.scipy.org/doc/scipy/dev/toolchain.html) and it looks like we shouldn't be using optional features like this.

Hmm, I'm somewhat confused by the toolchain roadmap... I'll look at the `calloc` implementation.